dynamic_programming/catalan_numbers.py

```diff
@@ -0,0 +1,100 @@
+"""
+Print all the Catalan numbers from 0 to n, n being the user input.
+
+ * The Catalan numbers are a sequence of positive integers that
+ * appear in many counting problems in combinatorics [1]. Such
+ * problems include counting [2]:
+ * - The number of Dyck words of length 2n
+ * - The number well-formed expressions with n pairs of parentheses
+ *   (e.g., `()()` is valid but `())(` is not)
+ * - The number of different ways n + 1 factors can be completely
+ *   parenthesized (e.g., for n = 2, C(n) = 2 and (ab)c and a(bc)
+ *   are the two valid ways to parenthesize.
+ * - The number of full binary trees with n + 1 leaves
+
+ * A Catalan number satisfies the following recurrence relation
+ * which we will use in this algorithm [1].
+ * C(0) = C(1) = 1
+ * C(n) = sum(C(i).C(n-i-1)), from i = 0 to n-1
+
+ * In addition, the n-th Catalan number can be calculated using
+ * the closed form formula below [1]:
+ * C(n) = (1 / (n + 1)) * (2n choose n)
+
+ * Sources:
+ *  [1] https://brilliant.org/wiki/catalan-numbers/
+ *  [2] https://en.wikipedia.org/wiki/Catalan_number
+"""
+
+
+class Catalan:
+    def __init__(self, upper_limit: int) -> None:
+        """
+        Create an object with a list of Catalan
+        numbers from 0 through upper_limit
+        """
+        self.catalan_numbers = [0] * (upper_limit + 1)
+        if upper_limit:
+            # Base case: C(0) = C(1) = 1
+            self.catalan_numbers[0] = 1
+            self.catalan_numbers[1] = 1
+
+            # Recurrence relation: C(i) = sum(C(j).C(i-j-1)), from j = 0 to i
+            for i in range(2, upper_limit + 1):
+                for j in range(i):
+                    self.catalan_numbers[i] += (
+                        self.catalan_numbers[j] * self.catalan_numbers[i - j - 1]
+                    )
+        elif upper_limit == 0:
+            self.catalan_numbers[0] = 1
+        print(self.catalan_numbers)
```

Comment:
I would avoid print statements in the algorithm itself as they produce unwanted side-effects.